### PR TITLE
feat: support jenkins as runtime context

### PR DESCRIPTION
**Pull request by mdelapenya**
- feat: add Jenkins as runtime executor
- chore: extract Jenkins context
- chore: extract gitlab and github context
- fix: return branch for github's branches
- chore: simplify scm contexts
- chore: support reading SCM from local runtime
- docs: add missing docs
- chore: rename method and change signature to return an struct
- chore: return nil SCM context

## What does this PR do?
This PR is adding the concept of SCM context, which means the execution context where the SCM is evaluated. It could be in a Github action, in a Gitlab runner, in a Jenkins pipeline, or in the local workspace.

For that, we are defining the variables that mark each context, so if present, that context is retrieved reading the environment variables defined for each of them on their docs.

- For Github: https://docs.github.com/en/actions/learn-github-actions/environment-variables#default-environment-variables
- For Gitlab: https://docs.gitlab.com/ee/ci/variables/predefined_variables.html
- For Jenkins: https://www.jenkins.io/doc/book/pipeline/getting-started/#global-variable-reference and https://www.jenkins.io/doc/book/pipeline/multibranch/#additional-environment-variables
- For Local execution: BRANCH and TARGET_BRANCH defines the SCM context. While BRANCH is mandatory, TARGET_BRANCH is optional and holds the name of the target branch a change request is sent.

## Why is it important?
We detected the need of adding Jenkins support for one reason: it could be the case where a developer does not own the Jenkins instance where the Jenkins OTEL plugin is configured, so they do not have control on how to use it.

Instead, they will be able to consume this CLI in the pipeline (using the binary, or the Docker image) as demonstrated in https://github.com/elastic/apm-pipeline-library/pull/1543, more specifically in https://github.com/elastic/apm-pipeline-library/pull/1543/files#diff-67ee4c112875e4f4586875b92aec928ff0a6f6c1cfeff2e1c08410d4be86bdacR38-R49

### CODE REVIEW

File: git.go [hunk 1]
```diff
@@ -82,31 +88,36 @@ func (scm *GitScm) calculateCommits() (*object.Commit, *object.Commit, error) {
...
 }
 
-// checkGitProvider identies the head sha and target branch from the environment variables that are
+// checkGiContext identifies the head sha and target branch from the environment variables that are
 // populated from a Git provider, such as Github or Gitlab. If no proprietary env vars are set, then it will
 // look up this tool-specific variable for the target branch.
-func checkGitProvider() (string, string, string, bool) {
+func checkGiContext() *ScmContext {
+	// in local branches, we are not in pull/merge requests
+	localContext := FromLocal()
```
Comment (mdelapenya) [inline git.go:96]: We prioritise the existence of the tool's own variables to be read first. This is allowing, i.e., running tests for each SCM context using Github actions

File: git.go [hunk 2]
```diff
@@ -140,11 +151,8 @@ func (scm *GitScm) contributeAttributes() []attribute.KeyValue {
 	}
 	gitAttributes = append(gitAttributes, attribute.Key(ScmRepository).StringSlice(origin.Config().URLs))
 
-	branch, err := scm.repository.Head()
-	if err != nil {
-		return gitAttributes
-	}
-	gitAttributes = append(gitAttributes, attribute.Key(ScmBranch).String(branch.Name().String()))
+	// do not read HEAD, and simply use the branch name coming from the SCM struct
```
Comment (mdelapenya) [inline git.go:154]: In the past, it added `HEAD` as the value of the `scm.branch` attribute

File: git_test.go [hunk 3]
```diff
@@ -207,153 +207,217 @@ func (r *FakeGitRepo) read() *GitScm {
...
 func TestGit_ContributeAttributesCloneOptions(t *testing.T) {
-	os.Setenv("TARGET_BRANCH", "master") // master branch is the base branch for the fake repository (octocat/hello-world)
-	defer os.Unsetenv("TARGET_BRANCH")
+	os.Setenv("BRANCH", "master") // master branch is the base branch for the fake repository (octocat/hello-world)
+	defer func() {
+		os.Unsetenv("BRANCH")
+	}()
 
-	scm := NewFakeGitRepo(t, WithCloneOptions(CloneOptionsRequest{Depth: 1})).withBranch("this-is-a-test-branch").addingFile("TEST-sample2.xml").removingFile("README").withCommit("This is a test commit").read()
+	scm := NewFakeGitRepo(t, WithCloneOptions(CloneOptionsRequest{Depth: 1})).read()
```
Comment (mdelapenya) [inline git_test.go:420]: This is a simplification, as there is no need to add commits to calculate the clone options

File: scm.go [hunk 4]
```diff
@@ -9,6 +9,129 @@ type Scm interface {
...
+// in their docs
+func FromJenkins() *ScmContext {
+	if os.Getenv("JENKINS_URL") == "" {
+		return nil
+	}
+
+	isPR := os.Getenv("CHANGE_ID") != ""  // only present on multibranch pipelines on Jenkins
+	headRef := os.Getenv("BRANCH_NAME")   // only present on multibranch pipelines on Jenkins
+	sha := os.Getenv("GIT_COMMIT")        // only present on multibranch pipelines on Jenkins
+	baseRef := os.Getenv("CHANGE_TARGET") // only present on multibranch pipelines on Jenkins
```
Comment (mdelapenya) [inline scm.go:93]: These variables are used by multibranch pipelines

**Review by kuisathaverat — APPROVED** [review]

**Pull request merged**